piptools/resolver.py

```diff
@@ -289,6 +291,15 @@ def _iter_dependencies(self, ireq):
         dependency_strings = self.dependency_cache[ireq]
         log.debug('  {:25} requires {}'.format(format_requirement(ireq),
                                                ', '.join(sorted(dependency_strings, key=lambda s: s.lower())) or '-'))
+
+        if self.exclude:
+            dep_set = {str(self.get_best_match(InstallRequirement.from_line(e))) for e in dependency_strings}
+            exclude_set = {str(self.get_best_match(InstallRequirement.from_line(e))) for e in self.exclude}
+            log.debug('Excluding following dependencies:')
+            for exclude in (dep_set & exclude_set):
+                log.debug('  {}'.format(exclude))
+            dependency_strings = dep_set - exclude_set
```

Comment:
By doing this, the dependencies becomes strictly pinned version to the best current candidate, which will make the resolving fail if there's a need of changing candidate.

Try the following requirements.in:
```
docker-compose
requests
```
It will fail with something like `Could not find a version that matches requests==2.11.1,==2.18.4`.